hydroflow/src/util/cli.rs

```diff
@@ -151,13 +171,74 @@ impl ConnectionPipe {
             ConnectionPipe::TcpPort(addr) => {
                 let stream = TcpStream::connect(addr).await.unwrap();
                 let (sink, source) = tcp_bytes(stream);
-                (Some(Box::pin(source)), Some(Box::pin(sink)))
+                ConnectedBidi {
+                    source: Some(Box::pin(source)),
+                    sink: Some(Box::pin(sink)),
+                }
+            }
+            ConnectionPipe::Bind(bound) => {
+                let bound = bound.accept().await;
+                ConnectedBidi {
+                    source: Some(bound.0),
+                    sink: Some(bound.1),
+                }
+            }
+            ConnectionPipe::Demux(_) => panic!("Cannot connect to a demux pipe directly"),
+        }
+    }
+}
+
+pub struct ConnectedDemux<T: Connected> {
+    pub sink: Option<Pin<Box<dyn Sink<(u32, T::Input), Error = io::Error> + Send + Sync>>>,
+}
+
+#[async_trait]
+impl<T: Connected> Connected for ConnectedDemux<T>
+where
+    <T as Connected>::Input: 'static + Sync,
+{
+    type Input = (u32, T::Input);
+
+    fn take_source(&mut self) -> Pin<Box<dyn Stream<Item = Result<BytesMut, io::Error>> + Send>> {
+        panic!("Cannot take source from a demux pipe");
+    }
+
+    fn take_sink(&mut self) -> Pin<Box<dyn Sink<Self::Input, Error = io::Error> + Send + Sync>> {
+        self.sink.take().unwrap()
+    }
+
+    async fn from_pipe(pipe: ConnectionPipe) -> Self {
+        match pipe {
+            ConnectionPipe::Demux(demux) => {
+                let mut connected_demux: HashMap<
+                    u32,
+                    Pin<Box<dyn Sink<T::Input, Error = io::Error> + Send + Sync>>,
+                > = HashMap::new();
+                for (id, pipe) in demux {
+                    connected_demux.insert(id, T::from_pipe(pipe).await.take_sink());
+                }
+
+                let unfold = sink::unfold(
+                    connected_demux,
+                    move |mut connected_demux, v: (u32, T::Input)| async move {
```

Comment:
@MingweiSamuel @zzlk would love your feedback on if there are better ways to do this. This seems not optimal but trying to implement my own `Sink` ended up being way too hard.

Reply:
should be fixed now!